PR: Documentation updates (#530)
spyder-ide/qtpy: README.md

[![pypi version](https://img.shields.io/pypi/v/qtpy.svg)](https://pypi.org/project/QtPy/)
[![conda version](https://img.shields.io/conda/vn/conda-forge/qtpy.svg)](https://www.anaconda.com/download/)
[![download count](https://img.shields.io/conda/dn/conda-forge/qtpy.svg)](https://www.anaconda.com/download/)

> [![OpenCollective Backers](https://opencollective.com/spyder/backers/badge.svg?color=blue)](https://opencollective.com/spyder/)

Reviewer: I think here what we want is to point to the `#sponsors` section bellow

Author: Done.

[![OpenCollective Backers](https://opencollective.com/spyder/backers/badge.svg?color=blue)](https://opencollective.com/spyder/)
[![Join the chat at https://gitter.im/spyder-ide/public](https://badges.gitter.im/spyder-ide/spyder.svg)](https://gitter.im/spyder-ide/public)<br>
[![PyPI status](https://img.shields.io/pypi/status/qtpy.svg)](https://github.com/spyder-ide/qtpy)
[![Github build status](https://github.com/spyder-ide/qtpy/workflows/Tests/badge.svg)](https://github.com/spyder-ide/qtpy/actions)
[![Coverage Status](https://coveralls.io/repos/github/spyder-ide/qtpy/badge.svg?branch=master)](https://coveralls.io/github/spyder-ide/qtpy?branch=master)

*Copyright © 2009– The Spyder Development Team*

## Description

**QtPy** is a small abstraction layer that lets you
write applications using a single API call to either PyQt or PySide.

It provides support for PyQt5, PySide2, PyQt6 and PySide6 using the Qt5 layout
(where the QtGui module has been split into QtGui and QtWidgets).

Basically, you can write your code as if you were using PyQt or PySide directly,
but import Qt modules from `qtpy` instead of `PyQt5`, `PySide2`, `PyQt6` or `PySide6`.

> Accordingly, when porting code between different Qt bindings (PyQt vs PySide) or Qt versions (Qt5 vs Qt6), QtPy makes this much more painless, and allows you to easily and incrementally transition between them. QtPy handles incompatibilities and differences between bindings or Qt versions for you while keeping your project running, so you can focus more on your own code and less on keeping track of supporting every Qt version and binding. Furthermore, when you do want to upgrade or support new bindings, it allows you to update your project module by module rather than all at once.  You can check out examples of this approach in projects using QtPy, like [git-cola](https://github.com/git-cola/git-cola/issues/232).

Reviewer: Probably the missing section related with success stories could include naming the `git-cola` project PR and also Spyder

Author: Done, I left the rest in the Features section.

```

## Installation

```bash
pip install qtpy
```

or

```bash
conda install qtpy
```

### Requirements

The installation requires one of the supported packages (PySide2, PySide6, PyQt5, PyQt6) as QtPy does not install any
binding by itself.
If several of these packages are found, PyQt5 is used by default.

> To set a specific binding, set the `QT_API` environment variable to on of the following values:

Reviewer: There is a typo here `on` -> `one`

Author: Fixed.

* `pyside2` (to use PySide2).
* `pyside6` (to use PySide6).
* `pyqt5` (to use PyQt5).

> * `pyqt6` (to use PyQt6).

Reviewer: Here the example code from the `__init__` file is missing. So we could add something like:

_For example if you want to use PyQt6 you could do:_

```python
import os
os.environ['QT_API'] = 'pyqt6'
from qtpy import QtGui, QtWidgets, QtCore
print(QtWidgets.QWidget)
```

Author: Done, I added a Bindings section.

## Description

**QtPy** provides a uniform interface to support PyQt5, PySide2, PyQt6, and PySide6 through a single codebase.
It abstracts the differences between bindings and versions, allowing software to remain portable and easier to maintain.

Import from `qtpy` instead of PySide/PyQt:

```python

> from qtpy import QtWidgets, QtCore

Reviewer 2: ```suggestion
from qtpy import QtCore, QtWidgets
```

Fix the import ordering to follow standard convention

app = QtWidgets.QApplication()
widget = QtWidgets.QWidget()
widget.show()
app.exec()
```

## Installation

> ```bash

Reviewer 2: ~~~suggestion
```shell
~~~

(Genericize this from the original, since we're touching it anyway)

pip install qtpy
```

or

> ```bash

Reviewer 2: ~~~suggestion
```shell
~~~

(Same here)

conda install qtpy
```

### Requirements

The installation requires one of the supported packages (PyQt5, PyQt6, PySide2, PySide6) as QtPy does not install any binding by itself.

> If several of these packages are found, PyQt5 is used by default.

Reviewer 2: ```suggestion
If multiple of these libraries are found, PyQt5 is used by default.
```

Use more appropriate diction and avoid word repetition

To set a specific binding, see [Bindings](#Bindings).

## Features

* Supports multiple Qt bindings (PyQt5/6, PySide2/6) without requiring conditional imports or logic branching.
* Detects and loads the available Qt binding automatically based on what is installed or already imported.

> * Normalizes the module structure to follow the modern Qt5 layout (`QtGui` / `QtWidgets`).

Reviewer 2: ```suggestion
* Normalizes the module structure to follow the Qt5 layout (`QtGui` / `QtWidgets`).
```

Qt 5 isn't so "modern" anymore (this was originally written when QtPy supported Qt 4 and Qt 5)

* Simplifies the process of porting applications between Qt5 and Qt6 by handling API incompatibilities internally.
* Enables incremental updates to project modules rather than requiring a full-scale rewrite when changing Qt providers.

### Bindings

To set a specific binding, set the `QT_API` environment variable to one of the following values:

| Value     | Binding |
|-----------|---------|
| `pyside6` | PySide6 |
| `pyside2` | PySide2 |
| `pyqt6`   | PyQt6   |
| `pyqt5`   | PyQt5   |

For example, to use PyQt6:
```python
import os
os.environ['QT_API'] = 'pyqt6'

> from qtpy import QtGui, QtWidgets, QtCore

Reviewer 2: ```suggestion
from qtpy import QtCore, QtGui, QtWidgets
```

Import order

| `qtpy.QT5`            | `True`/`False` if Qt5 is currently used.                    |
| `qtpy.QT6`            | `True`/`False` if Qt6 is currently used.                    |

### Compat Module

The `qtpy.compat` module provides wrappers for `QFileDialog` static methods and SIP/Shiboken functions.

| Source                               | Wrappers                           |
|--------------------------------------|------------------------------------|
| `QFileDialog.getExistingDirectory`   | `qtpy.compat.getexistingdirectory` |
| `QFileDialog.getOpenFileName`        | `qtpy.compat.getopenfilename`      |
| `QFileDialog.getOpenFileNames`       | `qtpy.compat.getopenfilenames`     |
| `QFileDialog.getSaveFileName`        | `qtpy.compat.getsavefilename`      |
| `sip.isdeleted` / `shiboken.isValid` | `qtpy.compat.isalive`              |

### Type Checker Integration

Type checkers have no knowledge of installed packages, so these tools require additional configuration.

A Command Line Interface (CLI) is offered to help with usage of QtPy (to get MyPy and Pyright/Pylance args/configurations).

#### Mypy

> The `mypy-args` command helps you to generate command line arguments for Mypy that will enable it to process the QtPy source files with the same API as QtPy itself would have selected.

Reviewer 2: ```suggestion
The `mypy-args` command generates command line arguments for Mypy that will enable it to process the QtPy source files with the same API as QtPy itself would have selected.
```

The original is rather verbose; we may as well be more direct here

To output a string of Mypy CLI args that will reflect the currently selected Qt API run:

> ```bash

Reviewer 2: ```suggestion
```shell
```

Another opportunistic tweak

qtpy mypy-args
```

For example, in an environment where PyQt5 is installed and selected (or the default fallback, if no binding can be found in the environment), this would output the following:

```text
--always-true=PYQT5 --always-false=PYSIDE2 --always-false=PYQT6 --always-false=PYSIDE6
```

Using Bash or a similar shell, this can be injected into the Mypy command line invocation:

```bash
mypy --package mypackage $(qtpy mypy-args)
```

#### Pyright/Pylance

> In the case of Pyright, instead of runtime arguments, it is required to create a config file for the project, called `pyrightconfig.json` or a `pyright` section in `pyproject.toml`. 

Reviewer 2: ```suggestion
In the case of Pyright, instead of runtime arguments you need to create a `pyrightconfig.json` config file for the project or a `pyright` section in `pyproject.toml`. 
```

The original had rather awkward unidiomatic grammar here

> Refer to the [Pyright Configuration](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) for reference.

Reviewer 2: ```suggestion
Refer to the [Pyright Configuration](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) for a full reference.
```

Make this a complete sentence

In order to set this configuration, QtPy offers the `pyright-config` command for guidance.

> To print the necessary configs for your project files run:

Reviewer 2: ```suggestion
To print the necessary configuration for your `pyrightconfig.json` or `pyproject.toml`, run:
```

Say this more clearly and explicitly, and ensure the next line makes more sense

```bash
qtpy pyright-config
```

If you don't have any, it is recommended to create them.

> For example, in an environment where PyQt5 is installed and selected (or the default fallback, if no binding can be found in the environment), this would output the following:

Reviewer 2: ```suggestion
If you don't have either of these, you should create them.

For example, in an environment where PyQt5 is installed and selected (or the default fallback, if no binding can be found in the environment), `qtpy pyright-config` would output the following:
```

* Fix the grammar and clarify the object of the first sentence
* Add a paragraph break since the "example" refers to an example of the usage of the `pyright-config` command generally, rather than the object of the previous sentence (creating those config files)
* Be explicit about what outputs the following, since again we now have some distance between the mention of the command and this example

`pyrightconfig.json`

```json
{
  "defineConstant": {
    "PYQT5": true,
    "PYSIDE2": false,
    "PYQT6": false,
    "PYSIDE6": false
  }
}
```

`pyproject.toml`

```toml
[tool.pyright.defineConstant]
PYQT5 = true
PYSIDE2 = false
PYQT6 = false
PYSIDE6 = false
```

> **Note**: These configurations are necessary for the correct usage of the default VSCode's type checking feature while using QtPy in your source code.

Reviewer 2: ```suggestion
**Note**: This configuration is necessary for the correct usage of the default VSCode type checking feature when using QtPy in your source code.
```

Fix multiple grammar errors from the original sentance

## Testing Matrix

> Currently, QtPy runs tests for different bindings on Linux, Windows and macOS, using Python 3.9, 3.11 and 3.13, and installing those bindings with `conda` and `pip`.

Reviewer 2: ```suggestion
Currently, QtPy runs tests for different bindings on Linux, Windows and macOS, using Python 3.9, 3.11 and 3.13, and installed via `conda` and `pip`.
```

Clarify phrasing from the original (it sounded like it was testing specifically the installation process itself rather than the package _as installed_ under either package manager.

For the PyQt bindings, the installation of extra packages is also checked via `pip`.

> Following this, the current test matrix looks something like this:

Reviewer 2: ```suggestion
The current test matrix looks something like this:
```

"Following this" isn't really necessary/makes sense here following your changes.

In the case of Pyright, instead of runtime arguments, it is required to create a
config file for the project, called `pyrightconfig.json` or a `pyright` section
in `pyproject.toml`. Refer to the
[Pyright Configuration](https://github.com/microsoft/pyright/blob/main/docs/configuration.md)
for reference. In order to set this configuration, QtPy offers the `pyright-config`
command for guidance.

> To add necessary configs in your project files run:

Reviewer: This command does not generate any files and rather outputs to the terminal what it was before in this section of the README, so efectivly:

```text
pyrightconfig.json:
{"defineConstant": {"PYQT5": true, "PYSIDE2": false, "PYQT6": false, "PYSIDE6": false}}

pyproject.toml:
[tool.pyright.defineConstant]
PYQT5 = true
PYSIDE2 = false
PYQT6 = false
PYSIDE6 = false
```

<img width="730" height="168" alt="Image" src="https://github.com/user-attachments/assets/db9a318b-1e93-41d8-ad05-d57b856ac979" />

Author: Done, I clarified that it's just a print.

Reviewer: Seems like there is still some text bellow this part explaining things as if files would be generated: https://github.com/spyder-ide/qtpy/pull/530/changes#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R180-R197

Author: The output for the two files is included, I just wrapped it in the proper formatting per file. I hope I'm not missing something obvious haha
https://github.com/spyder-ide/qtpy/pull/530/changes

If I'm missing anything, please feel free to just edit it, I'm totally fine with that. This is your repo :)

Reviewer: Maybe a little nitpick from my side but the phrase bellow _`this would output the following files:`_ still sounds a little bit missleading to me :thinking: Anyhow added a commit taking care of that :+1:

* `PyQtDataVisualization` and `PyQt6-DataVisualization`
* `PyQtNetworkAuth` and `PyQt6-NetworkAuth`
* `PyQtPurchasing`
* `PyQtWebEngine` and `PyQt6-WebEngine`
* `QScintilla` and `PyQt6-QScintilla`

## Attribution and Acknowledgments

This project is based on the [pyqode.qt](https://github.com/pyQode/pyqode.qt)
project and the [spyderlib.qt](https://github.com/spyder-ide/spyder/tree/2.3/spyderlib/qt)
module from the [Spyder](https://github.com/spyder-ide/spyder) project, and
also includes contributions adapted from
[qt-helpers](https://github.com/glue-viz/qt-helpers), developed as part of the
[glue](http://glueviz.org) project.

Unlike `pyqode.qt` this is not a namespace package, so it is not tied
to a particular project or namespace.

## License

> This project is released under the [MIT license](LICENSE).

Reviewer: This will need to change depending on the decision we make about the changes to the license (no extension vs `.txt` extension)

Author: I reverted the commit.

* `PyQtDataVisualization` and `PyQt6-DataVisualization`
* `PyQtNetworkAuth` and `PyQt6-NetworkAuth`
* `PyQtPurchasing`
* `PyQtWebEngine` and `PyQt6-WebEngine`
* `QScintilla` and `PyQt6-QScintilla`

## Attribution and Acknowledgments

> This project is based on the [pyqode.qt](https://github.com/pyQode/pyqode.qt) project and the [spyderlib.qt](https://github.com/spyder-ide/spyder/tree/2.3/spyderlib/qt) module from the [Spyder](https://github.com/spyder-ide/spyder) project, and also includes contributions adapted from [qt-helpers (https://github.com/glue-viz/qt-helpers), developed as part of the [glue](http://glueviz.org) project.

Reviewer 2: ```suggestion
This project is based on the [pyqode.qt](https://github.com/pyQode/pyqode.qt) project and the [spyderlib.qt](https://github.com/spyder-ide/spyder/tree/2.3/spyderlib/qt) module from the [Spyder](https://github.com/spyder-ide/spyder) project, and also includes contributions adapted from [qt-helpers](https://github.com/glue-viz/qt-helpers), developed as part of the [glue](http://glueviz.org) project.
```

Fix Markdown link syntax mistake

Unlike `pyqode.qt` this is not a namespace package, so it is not tied to a particular project or namespace.

## Success Stories

> You can check out examples of how QtPy adds support for multiple bindings and allows to incrementally update packages here:

Reviewer 2: ```suggestion
You can check out examples of how QtPy adds support for multiple bindings and allows for incremental updates to new binding versions here:
```

Clarify the phrasing here some

* [git-cola](https://github.com/git-cola/git-cola/issues/232)
* [spyder](https://github.com/spyder-ide/spyder)

## License

This project is released under the [MIT license](LICENSE.txt).

## Contributing

> Everyone is welcome to contribute! See our [Contributing guide](CONTRIBUTING.md) for more details.

Reviewer 2: ```suggestion
Everyone is welcome to contribute!
See our [Contributing guide](CONTRIBUTING.md) for more details.
```

Separate sentences (possibly whatever regex you used to do the splitting didn't consider the `!`)

## Sponsors

> QtPy is funded thanks to the generous support of [![Chan Zuckerberg Initiative](https://raw.githubusercontent.com/spyder-ide/spyder/master/img_src/czi.png)](https://chanzuckerberg.com/)[![Numfocus](https://i2.wp.com/numfocus.org/wp-content/uploads/2017/07/NumFocus_LRG.png?fit=320%2C148&ssl=1)](https://numfocus.org/) and the donations we have received from our users around the world through [Open Collective](https://opencollective.com/spyder/):

Reviewer 2: ```suggestion
QtPy is funded thanks to the generous support of our users from around the world through [Open Collective](https://opencollective.com/spyder/) and [NumFOCUS](https://numfocus.org/project/spyder):
```

Update this as CZI does not currently directly fund QtPy, and right now we don't have a direct NumFOCUS general funding for Spyder as opposed to people funding it through NumFOCUS, and also fix the broken layout.
